File: ios/Classes/GattGroupRequest.swift
import Flutter
import XyBleSdk
import CoreBluetooth
import SwiftProtobuf
import Promises

/// Handles a set of calls for one device in one connection
struct GattGroupRequest: GattRequestHandler {

    static func process(arguments: Any?, result: @escaping FlutterResult) {
        guard
            let operations = self.operations(arguments),
            let device = self.device(operations.deviceID)
            else {
                result(nil)
                return
        }

        self.execute(device, operations: operations).then { responses in
            self.relayResponse(GattResponseList.with { $0.responses = responses }, result: result)
        }.catch { error in
            // do something
        }
    }

    static func execute(_ device: XYBluetoothDevice?, operations: GattOperationList) -> Promise<[GattResponse]> {
        guard let device = device else {
            return Promise([])
        }

        let operationPromise = Promise<[GattResponse]>.pending()

        var responses = [GattResponse]()
        var results = [XYBluetoothResult]()

        device.connection {
            results = operations.operations.map { self.runCall(for: device, operation: $0) }
        }.then {
            // Assemble the responses
            for (index, operation) in operations.operations.enumerated() {
                guard let bleResult = results[safe: index] else {
                    // do something
                    continue
                }
                if bleResult.hasError {
                    responses.append(self.response(operation, error: bleResult.error!))
                } else {
                    responses.append(self.response(operation, response: bleResult))
                }
            }

            operationPromise.fulfill(responses)

        }.catch { error in
            // Do something here
            operationPromise.reject(error)
        }.always {
            self.cleanup(for: device, operations: operations)
        }

        return operationPromise
    }

}

File: android/src/main/kotlin/network/xyo/ble/xyo_ble/GattGroupRequest.kt
package network.xyo.ble.xyo_ble

import io.flutter.plugin.common.MethodChannel
import network.xyo.ble.flutter.protobuf.Gatt
import network.xyo.ble.generic.gatt.peripheral.XYBluetoothResult
import network.xyo.ble.generic.scanner.XYSmartScan

@kotlin.ExperimentalUnsignedTypes
class GattGroupRequest: GattRequestHandler() {

    companion object {
        suspend fun process(smartScan: XYSmartScan, arguments: Any?, result: MethodChannel.Result): Boolean {
                val operations = operations(arguments) ?: return false
                val responses = execute(smartScan, operations) ?: return false
                val responseListBuilder = Gatt.GattResponseList.newBuilder()
                for (response: Gatt.GattResponse in responses) {
                    responseListBuilder.addResponses(response)
                }
                relayResponse(responseListBuilder.build(), result)
                return true
        }

        suspend fun execute(smartScan: XYSmartScan, operations: Gatt.GattOperationList): MutableList<Gatt.GattResponse>? {
            val responses = mutableListOf<Gatt.GattResponse>()
            for (operation in operations.operationsList) {
                val device = smartScan.devices[operation.deviceId] ?: return null
                device.connection {
                    val result = runCall(device, operation)
                    responses.add(response(operation, result))
                    return@connection XYBluetoothResult(true)
                }
            }
            return responses
        }
    }

}
